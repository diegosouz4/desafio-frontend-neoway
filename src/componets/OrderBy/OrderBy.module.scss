.wrapper {
  width: 92%;
  max-width: 1200px;
  margin: 0 auto;
  margin-bottom: 0.3125rem;
  fieldset {
    border: 0;
    box-sizing: border-box;
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: flex-end;
    gap: 0.625rem;

    position: relative;
    animation: fadeInUp 1s forwards ease;
    opacity: 0;
    transform: translate3d(0, 50px, 0);

    .label {
      color: var(--color-white-100);
      display: block;
      font-size: 0.75rem;
    }

    .select {
      color: var(--color-text);
      border: 0;
      border-radius: 0.25rem;
      border: 2px solid transparent;
      background-color: #cbcbcb;
      height: 2rem;
      padding: 0.3125rem 0.625rem;
      width: 100%;
      max-width: 6.5rem;

      &:focus,
      &:hover {
        border: 2px solid var(--color-main);
        outline-color: var(--color-main);
      }
    }
  }
}

@keyframes fadeInUp {
  from {
    opacity: 0;
    transform: translate3d(0, 50px, 0);
  }
  to {
    opacity: 1;
    transform: translate3d(0, 0, 0);
  }
}
