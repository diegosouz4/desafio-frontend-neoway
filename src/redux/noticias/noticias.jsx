import { createSlice } from "@reduxjs/toolkit";

const initialState = {
  news: {
    loading: false,
    data: null,
    error: null,
  },
};

const slice = createSlice({
  name: "noticias",
  initialState: initialState,
  reducers: {
    fetchNewsStart(state) {
      (state.news.loading = true),
        (state.news.data = null),
        (state.news.error = null);
    },
    fetchNewsSuccess(state, action) {
      (state.news.loading = false),
        (state.news.data = action.payload),
        (state.news.error = null);
    },
    fetchNewsError(state, action) {
      (state.news.loading = false),
        (state.news.data = null),
        (state.news.error = action.payload);
    },
  },
});

const { fetchNewsStart, fetchNewsSuccess, fetchNewsError } = slice.actions;

export const fetchNoticias = (query) => async (dispatch) => {
  const { q, language, sortBy } = query;
  const url = `https://newsapi.org/v2/everything?q=${q}&language=${language}&sortBy=${sortBy}&from=2023-09-24&apiKey=${
    import.meta.env.VITE_API_TOKEN_TESTE
  }&pageSize=100`;

  try {
    dispatch(fetchNewsStart());
    const response = await fetch(url);
    const data = await response.json();
    //Removendo as notícias
    data.articles = await data.articles.filter(
      (item) => item.title !== "[Removed]" && item.title !== null
    );
    //Alterando os id's null pelo index da notícia.
    data.articles = await data.articles.map((item, index) =>({...item, source: {...item.source, id: index}}));
    return dispatch(fetchNewsSuccess(data));
  } catch (error) {
    dispatch(fetchNewsError(error.message));
  }
};

export default slice.reducer;
