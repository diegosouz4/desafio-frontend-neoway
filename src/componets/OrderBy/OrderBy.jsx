import React from "react";
import style from "./OrderBy.module.scss";
import { useDispatch, useSelector } from "react-redux";
import { updateNewsByOrder } from "../../redux/noticias/noticias";
import PropTypes from "prop-types";

export default function OrderBy({ origin }) {
  const [sortBy, setSortBy] = React.useState("");
  const dispatch = useDispatch();
  const { news } = useSelector((rootReducer) => rootReducer.noticias);

  const handleChange = ({ target }) => {
    if (!target.value) return;

    setSortBy(target.value);
    if (!origin)
      return dispatch(updateNewsByOrder(news.data.articles, target.value));

    return dispatch(updateNewsByOrder(news.likes, target.value, origin));
  };

  return (
    <div className={style.wrapper}>
      <fieldset>
        <label className={style.label} htmlFor="orderby">
          Ordernar por:{" "}
        </label>
        <select
          className={style.select}
          id="orderby"
          value={sortBy}
          onChange={handleChange}
        >
          <option value="">Selecionar</option>
          <option value="author">Autor</option>
          <option value="publishedAt">Publicação</option>
          <option value="title">Título</option>
        </select>
      </fieldset>
    </div>
  );
}

OrderBy.propTypes = {
  origin: PropTypes.string,
};
