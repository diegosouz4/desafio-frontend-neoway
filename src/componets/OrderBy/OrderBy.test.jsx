import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { useSelector, Provider, useDispatch } from "react-redux";
import OrderBy from "./OrderBy";
import configureStore from "redux-mock-store";
import "@testing-library/jest-dom";
import { updateNewsByOrder } from "../../redux/noticias/noticias";

jest.mock("../../redux/noticias/noticias", () => {
  return { updateNewsByOrder: jest.fn() };
});

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

const mockStore = configureStore([]);

describe("<OrderBy />", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      noticias: {
        news: {
          data: {
            articles: [],
          },
          likes: [],
        },
      },
    });
  });

  it("should OrdernBy was rendered", () => {
    useSelector.mockReturnValue({
      noticias: {
        news: {
          data: {
            articles: [],
          },
          likes: [],
        },
      },
    });

    const { asFragment } = render(
      <BrowserRouter>
        <Provider store={store}>
          <OrderBy origin="likes" />
        </Provider>
      </BrowserRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should dispath the likes array orderby author", async () => {
    const dispatch = jest.fn();

    useDispatch.mockReturnValue(dispatch);

    useSelector.mockReturnValue({
      news: {
        data: {
          articles: [],
        },
        likes: [],
      },
    });

    render(
      <BrowserRouter>
        <Provider store={store}>
          <OrderBy origin="likes" />
        </Provider>
      </BrowserRouter>
    );

    const select = screen.getByLabelText("Ordernar por:");
    expect(select).toBeInTheDocument();

    await fireEvent.change(select, { target: { value: "author" } });

    expect(updateNewsByOrder).toHaveBeenCalledWith(
      store.getState().noticias.news.likes,
      "author",
      "likes"
    );
  });

  it("should dispath the articles array orderby author", async () => {
    const dispatch = jest.fn();

    useDispatch.mockReturnValue(dispatch);

    useSelector.mockReturnValue({
      news: {
        data: {
          articles: [],
        },
        likes: [],
      },
    });

    render(
      <BrowserRouter>
        <Provider store={store}>
          <OrderBy />
        </Provider>
      </BrowserRouter>
    );

    const select = screen.getByLabelText("Ordernar por:");
    expect(select).toBeInTheDocument();

    await fireEvent.change(select, { target: { value: "author" } });

    expect(updateNewsByOrder).toHaveBeenCalledWith(
      store.getState().noticias.news.data.articles,
      "author"
    );
  });
});
