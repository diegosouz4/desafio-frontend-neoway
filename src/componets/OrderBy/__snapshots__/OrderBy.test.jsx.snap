// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderBy /> should OrdernBy was rendered 1`] = `
<DocumentFragment>
  <div>
    <fieldset>
      <label
        for="orderby"
      >
        Ordernar por: 
      </label>
      <select
        id="orderby"
      >
        <option
          value=""
        >
          Selecionar
        </option>
        <option
          value="author"
        >
          Autor
        </option>
        <option
          value="publishedAt"
        >
          Publicação
        </option>
        <option
          value="title"
        >
          Título
        </option>
      </select>
    </fieldset>
  </div>
</DocumentFragment>
`;
